Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,320 @@
+  simp [coproduct]
+  rfl
+
+variable {A} in
+theorem preservesCoproductEval (F : C ⥤ A) {J : Type w} (f : J → FormalCoproduct.{w} C) :
+    PreservesColimit (Discrete.functor f) ((eval.{w} C A).obj F) := by
+  refine CategoryTheory.Limits.PreservesCoproduct.of_iso_comparison _ _ (i := ?_)
+  rw [sigmaComparison_eq, ← Functor.mapIso_inv, ← Iso.trans_inv, ← Iso.symm_hom]
+  infer_instance
+
+end HasCoproducts
+
+
+section HasProducts
+
+variable [HasProducts.{w} A]
+
+variable (C) in
+@[simps] noncomputable
+def evalOp : (Cᵒᵖ ⥤ A) ⥤ ((FormalCoproduct.{w} C)ᵒᵖ ⥤ A) where
+  obj F :=
+    { obj X := ∏ᶜ fun (i : X.unop.I) ↦ F.obj (op (X.unop.obj i))
+      map f := Pi.lift fun i ↦ Pi.π _ (f.unop.f i) ≫ F.map (f.unop.φ i).op }
+  map α := { app f := Pi.map fun i ↦ α.app (op (f.unop.obj i)) }
+
+variable {A} in
+noncomputable def evalOpOf :
+    evalOp C A ⋙ (whiskeringLeft _ _ A).obj (of C).op ≅ Functor.id (Cᵒᵖ ⥤ A) :=
+  NatIso.ofComponents fun F ↦ NatIso.ofComponents fun x ↦
+    ⟨Pi.π _ PUnit.unit, Pi.lift fun _ ↦ 𝟙 _, by aesop, by simp⟩
+
+variable {A} in
+@[simps!] noncomputable
+def evalOpCoproductIso (F : Cᵒᵖ ⥤ A) {J : Type w} (f : J → FormalCoproduct.{w} C) :
+    ((evalOp C A).obj F).obj (op (coproduct J f).cocone.pt) ≅
+      ∏ᶜ fun j ↦ ((evalOp C A).obj F).obj (op (f j)) :=
+  (piPiIso (fun j ↦ (f j).I) (fun j x ↦ F.obj (op ((f j).obj x)))).symm
+
+lemma evalOpCoproductIso_comp (F : Cᵒᵖ ⥤ A) {J : Type w} (f : J → FormalCoproduct.{w} C) (j : J) :
+    (evalOpCoproductIso F f).hom ≫ Pi.π _ j =
+      ((evalOp C A).obj F).map (op ((coproduct J f).cocone.ι.app ⟨j⟩)) :=
+  (Pi.lift_π _ _).trans (by simp [coproduct])
+
+lemma π_eq {J : Type w} (f : J → FormalCoproduct.{w} C) (j : J) :
+    Pi.π (op ∘ f) j = (opCoproductIsoProduct _).inv ≫ op (colimit.ι _ _) :=
+  (opCoproductIsoProduct_inv_comp_ι _ _).symm
+
+lemma piComparison_eq (F : Cᵒᵖ ⥤ A) {J : Type w} (f : J → FormalCoproduct.{w} C) :
+    piComparison ((evalOp C A).obj F) (op ∘ f) =
+      ((evalOp C A).obj F).map ((opCoproductIsoProduct _).inv ≫ (coproductIsoCoproduct _).op.inv) ≫
+        (evalOpCoproductIso F f).hom := by
+  refine Pi.hom_ext _ _ fun j ↦ Pi.hom_ext _ _ fun x ↦ ?_
+  rw [piComparison_comp_π, evalOp_obj_map, Pi.lift_π, evalOp_obj_map, evalOpCoproductIso_hom,
+    Category.assoc, Category.assoc]
+  conv => enter [2,2]; rw [← Category.assoc]; enter [1]; exact Pi.lift_π _ _
+  conv => enter [2,2]; exact Pi.lift_π _ _
+  refine Eq.symm ((Pi.lift_π _ _).trans ?_)
+  congr 1
+  · congr 3; simp [π_eq]
+  · congr 1; simp [π_eq]
+  · congr 1
+    · rw [π_eq]; rfl
+    · rw [π_eq]; rfl
+
+variable {A} in
+theorem preservesProductEvalOp (F : Cᵒᵖ ⥤ A) {J : Type w} (f : J → FormalCoproduct.{w} C) :
+    PreservesLimit (Discrete.functor (op ∘ f)) ((evalOp.{w} C A).obj F) := by
+  refine CategoryTheory.Limits.PreservesProduct.of_iso_comparison _ _ (i := ?_)
+  rw [piComparison_eq, ← Iso.trans_inv, ← Functor.mapIso_inv, ← Iso.symm_hom]
+  infer_instance
```

Comment:
We might get away with using the corresponding fact for colimits and using the existing lemmas translating limits to colimits in opposites, rather than redoing everything a second time?

Reply:
Yes, I feel like that is possible, but now the proof is short enough and it might take me more brain power to figure out how to correctly translate everything to the opposite categories...